### ISSUE TITLE
constantly getting a no target with given id found

### ISSUE THREAD

**Issue by choval** [chrome-php/headless-chromium-php#38]
This only happens when using the userData, I need to keep the environment instead of having an entire fresh instance created on every load.

```
[2018-09-29 00:03:49] DEBUG socket(1): ← receiving data:{"error":{"code":-32602,"message":"No target with given id found"},"id":22}
```

Which eventually breaks the Session constructor because the sessionId is not returned.

```
PHP Fatal error:  Uncaught TypeError: Argument 2 passed to HeadlessChromium\Communication\Session::__construct() must be of the type string, null given, called in .../vendor/chrome-php/chrome/src/Communication/Connection.php on line 240 and defined in .../vendor/chrome-php/chrome/src/Communication/Session.php:42
```

Current workaround:

```
src/Communication/Connection.php
commented out the return type of createSession so I can:

$sessionId = $response['result']['sessionId'] ?? false;
if(!$sessionId)
  return;

src/Browser.php
Not creating a target if no session is returned

if(!$session)
  return;
```

Not sure if it's the best approach as I just started to fiddle around with this library...
I will update once I track everything down.

Great work by the way!

**Comment by gsouf** [comment]
Hi @choval thanks for reporting. I tried to reproduce but couldnt get the error you mentioned with the following script:

```php

use \HeadlessChromium\BrowserFactory;

$browserFactory = new BrowserFactory('chromium-browser');

$browser = $browserFactory->createBrowser([
    'debugLogger'     => 'php://stdout', // will enable verbose mode
    'userDataDir'     => '/tmp/persistDataDir'
]);

$page = $browser->createPage();

$page->navigate('http://example.com')->waitForNavigation();
```

Can you please provide an isolated example where the error occurs?

**Comment by choval** [comment]
Took me a while, but I've managed to consistently reproduce the error.

```
OS: macOs 10.13.6 (High Sierra)
PHP: 7.1.21 (homebrew)

$ php -v
PHP 7.1.21 (cli) (built: Aug 20 2018 20:41:15) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.1.0, Copyright (c) 1998-2018 Zend Technologies
    with Zend OPcache v7.1.21, Copyright (c) 1999-2018, by Zend Technologies

$ uname -a
Darwin OsBook.home.osvaldojiang.com 17.7.0 Darwin Kernel Version 17.7.0: Thu Jun 21 22:53:14 PDT 2018; root:xnu-4570.71.2~1/RELEASE_X86_64 x86_64

$ composer info
apix/log               1.2.1  Minimalist, thin and fast PSR-3 compliant (multi-bucket) logger.
chrome-php/chrome      v0.2.3 Instrument headless chrome/chromium instances from php
evenement/evenement    v3.0.1 Événement is a very simple event dispatching library for PHP
psr/log                1.0.2  Common interface for logging libraries
symfony/filesystem     v4.1.6 Symfony Filesystem Component
symfony/polyfill-ctype v1.9.0 Symfony polyfill for ctype functions
symfony/process        v4.1.6 Symfony Process Component
wrench/wrench          v2.0.8 PHP WebSocket client/server library
```

This script duplicates the error:

```
<?php
// Autoload
require('vendor/autoload.php');

$bin = '"/Applications/Google Chrome.app/Contents/MacOS/Google Chrome"';

// Page that registers a serviceworker
$url = 'https://googlechrome.github.io/samples/service-worker/post-message/index.html';

use \HeadlessChromium\BrowserFactory;
$browserFactory = new BrowserFactory($bin);

$browser = $browserFactory->createBrowser([
    'debugLogger' => 'php://stdout',
//    'userDataDir' => '/tmp/persistDataDir',
]);

// Preload, this creates session data on the browser
$page = $browser->createPage();
$page->navigate($url)->waitForNavigation();
$page->close();

$page = $browser->createPage();
$page->navigate($url)->waitForNavigation();
$page->evaluate('var e=document;e.open(),e.write("<p>TEST</p>"),e.close();')->getReturnValue();
// The previous JS sends a new target and destroy on the same line
sleep(8);
// During the above sleep, the library tries to attach to the new target, which was destroyed in the same line

// Do not exit, and try to do something
$test = $page->evaluate('document.location.toString()')->getReturnValue();
var_dump($test);
```

Attached the debug log.
[log.txt](https://github.com/chrome-php/headless-chromium-php/files/2444114/log.txt)

```
[2018-10-03 21:50:57] DEBUG socket(1): ← receiving data:{"method":"Target.targetCreated","params":{"targetInfo":{"targetId":"533E257FCE7D806F44E12E4A39600478","type":"service_worker","title":"Service Worker https://googlechrome.github.io/samples/service-worker/post-message/service-worker.js","url":"https://googlechrome.github.io/samples/service-worker/post-message/service-worker.js","attached":false,"browserContextId":"C96369E40B37858B9CBF9B4909CB43B3"}}}
[2018-10-03 21:50:57] DEBUG socket(1): ← receiving data:{"method":"Target.targetDestroyed","params":{"targetId":"533E257FCE7D806F44E12E4A39600478"}}
...
[2018-10-03 21:50:57] DEBUG connection: ⇶ dispatching method:Target.targetCreated
[2018-10-03 21:50:57] DEBUG socket(1): → sending data:{"id":35,"method":"Target.attachToTarget","params":{"targetId":"533E257FCE7D806F44E12E4A39600478"}}
[2018-10-03 21:50:57] DEBUG connection: ⇶ dispatching method:Target.targetDestroyed
...
[2018-10-03 21:50:57] DEBUG socket(1): ← receiving data:{"error":{"code":-32602,"message":"No target with given id found"},"id":35}
```

The line that creates the issue:
`var e=document;e.open(),e.write("<p>TEST</p>"),e.close();`

The open creates a target, and the same line deletes the target. The library sees the new target, and tries to attach to it, but it's been deleted already.

**Comment by gsouf** [comment]
Amazing, I'll investigate, thanks @choval !

**Comment by gsouf** [comment]
@choval that's fixed in #38

Thanks for your help figuring out where the issue was coming from. Not an easy pick!

**Comment by choval** [comment]
@gsouf Amazing, thanks!
